uploader/creds.go

```diff
@@ -0,0 +1,140 @@
+package uploader
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/request"
+	securejoin "github.com/cyphar/filepath-securejoin"
+)
+
+// Emulates the EC2RoleCreds Provider but using a custom path
+// https://github.com/aws/aws-sdk-go/blob/master/aws/credentials/ec2rolecreds/ec2_role_provider.go
+
+const ProviderName = "IMDSRoleProvider"
+
+// IMDSRoleProvider will attempt to load credentials from the imds proxy in a special path
+type IMDSRoleProvider struct {
+	credentials.Expiry
+
+	Client *ec2metadata.EC2Metadata
+
+	ExpiryWindow time.Duration
+}
+
+func NewIMDSCredentials(c client.ConfigProvider) *credentials.Credentials {
+	p := &IMDSRoleProvider{
+		Client: ec2metadata.New(c),
+	}
+
+	return credentials.NewCredentials(p)
+}
+
+const credentialPath = "iam/logging/security-credentials/"
+
+// Retrieve retrieves credentials from the Instance Metadata Service.
+// Error will be returned if the request fails, or unable to extract the desired credentials.
+func (i *IMDSRoleProvider) Retrieve() (credentials.Value, error) {
+	return i.RetrieveWithContext(aws.BackgroundContext())
+}
+
+func (i *IMDSRoleProvider) RetrieveWithContext(ctx credentials.Context) (credentials.Value, error) {
+	credsList, err := requestCredList(ctx, i.Client)
+	if err != nil {
+		return credentials.Value{ProviderName: ProviderName}, err
+	}
+
+	if len(credsList) == 0 {
+		return credentials.Value{ProviderName: ProviderName}, awserr.New("EmptyEC2RoleList", "empty EC2 Role list", nil)
+	}
+	credsName := credsList[0]
+
+	roleCreds, err := requestCred(ctx, i.Client, credsName)
+	if err != nil {
+		return credentials.Value{ProviderName: ProviderName}, err
+	}
+
+	i.SetExpiration(roleCreds.Expiration, i.ExpiryWindow)
+
+	return credentials.Value{
+		AccessKeyID:     roleCreds.AccessKeyID,
+		SecretAccessKey: roleCreds.SecretAccessKey,
+		SessionToken:    roleCreds.Token,
+		ProviderName:    ProviderName,
+	}, nil
+}
+
+// A ec2RoleCredRespBody provides the shape for unmarshaling credential
+// request responses.
+type ec2RoleCredRespBody struct {
+	// Success State
+	Expiration      time.Time
+	AccessKeyID     string
+	SecretAccessKey string
+	Token           string
+
+	// Error state
+	Code    string
+	Message string
+}
+
+func requestCredList(ctx aws.Context, client *ec2metadata.EC2Metadata) ([]string, error) {
+	resp, err := client.GetMetadataWithContext(ctx, credentialPath)
```

Comment:
Any reason you have to implement your own AWS endpoint resolver, and you can't just supply an alternate endpoint to the one constructed by ec2metadatutils when you configure it? You should be able to have a custom endpoint resolver, and it will do all the renew / retry magic for you:
https://docs.aws.amazon.com/sdk-for-go/api/aws/#Config.WithEndpoint

Reply:
I'd love to not have to re-implement much of this.  I'm not sure I see how the endpoint would help here since the path that I'm overriding here was similarly a [const in the aws-sdk](https://github.com/aws/aws-sdk-go/blob/master/aws/credentials/ec2rolecreds/ec2_role_provider.go#L137).  It could very easily afford configuration but it does not currently.

Are you suggesting that we host a entire duplicate endpoint for the logging case?  I'm open to the idea just seems like there's more work to be done to do that instead of just providing the provider here.

Reply:
Hrrrm. That seems to have different behaviour than every other AWS resolver where it appends to the endpoint that the thing responds with.

Something like this should work though:
```
package main

import (
	"fmt"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
	"github.com/aws/aws-sdk-go/aws/request"
	"github.com/aws/aws-sdk-go/aws/session"
	"log"
	"os"
)
import "github.com/aws/aws-sdk-go/aws/ec2metadata"

func main() {
	log.SetOutput(os.Stderr)

	cfg := (&aws.Config{
	}).WithLogLevel(aws.LogDebugWithHTTPBody | aws.LogDebugWithRequestRetries | aws.LogDebugWithRequestErrors | aws.LogDebugWithEventStreamBody | aws.LogDebugWithSigning)
	s := session.Must(session.NewSessionWithOptions(session.Options{
		Config: *cfg,
		EC2IMDSEndpoint: "http://localhost:1234",
	}))
	s.Handlers.Build.PushBack(func(request *request.Request) {
		log.Println("original url: %s", request.HTTPRequest.URL.String())
		request.HTTPRequest.URL.Path = "/logging" + request.HTTPRequest.URL.Path
		log.Println("new url: %s", request.HTTPRequest.URL.String())
	})
	md := ec2metadata.New(s)
	creds := ec2rolecreds.NewCredentialsWithClient(md)
	val, err :=creds.Get()
	fmt.Println(val)
	fmt.Println(err)
}
```

See the output:
```
sargun:imds sargun$ go run .
2021/01/21 00:22:31 original url: %s http://localhost:1234/latest/meta-data/iam/security-credentials/
2021/01/21 00:22:31 new url: %s http://localhost:1234/logging/latest/meta-data/iam/security-credentials/
2021/01/21 00:22:31 original url: %s http://localhost:1234/latest/api/token
2021/01/21 00:22:31 new url: %s http://localhost:1234/logging/latest/api/token
2021/01/21 00:22:31 DEBUG: Request ec2metadata/GetToken Details:
---[ REQUEST POST-SIGN ]-----------------------------
PUT /logging/latest/api/token HTTP/1.1
Host: localhost:1234
User-Agent: aws-sdk-go/1.36.29 (go1.13.14; darwin; amd64)
Content-Length: 0
X-Aws-Ec2-Metadata-Token-Ttl-Seconds: 21600
Accept-Encoding: gzip


-----------------------------------------------------
2021/01/21 00:22:31 DEBUG: Send Request ec2metadata/GetToken failed, attempt 0/2, error RequestError: send request failed
caused by: Put http://localhost:1234/logging/latest/api/token: dial tcp [::1]:1234: connect: connection refused
2021/01/21 00:22:31 DEBUG: Retrying Request ec2metadata/GetToken, attempt 1
2021/01/21 00:22:31 DEBUG: Request ec2metadata/GetToken Details:
---[ REQUEST POST-SIGN ]-----------------------------
PUT /logging/latest/api/token HTTP/1.1
Host: localhost:1234
User-Agent: aws-sdk-go/1.36.29 (go1.13.14; darwin; amd64)
Content-Length: 0
X-Aws-Ec2-Metadata-Token-Ttl-Seconds: 21600
Accept-Encoding: gzip
```

I 
```

Reply:
You shouldn't rewrite the `/logging/latest/api/token` endpoint, so maybe a *tiny* bit more work required. But, yeah, I think the above approach should work of adding a hook in the session.

Alternatively, the user agent or headers trick means no rewriting required, just injection of the info.